src/linear_algebra/dual.lean

```diff
@@ -398,3 +407,171 @@ begin
 end
 
 end dual_pair
+
+namespace submodule
+
+universes u v w
+
+variables {R : Type u} {M : Type v} [comm_ring R] [add_comm_group M] [module R M]
+variable {W : submodule R M}
+
+/-- The `dual_annihilator` of a submodule `W` is the set of linear maps `φ` such
+  that `φ w = 0` for all `w ∈ W`. -/
+def dual_annihilator {R : Type u} {M : Type v} [comm_ring R] [add_comm_group M]
+  [module R M] (W : submodule R M) : submodule R $ module.dual R M :=
+{ carrier := { φ | ∀ w ∈ W, φ w = 0 },
+  zero_mem' := by simp,
+  add_mem' :=  by { intros φ ψ hφ hψ w hw,
+    rw [linear_map.add_apply, hφ w hw, hψ w hw, add_zero] },
+  smul_mem' := by { intros c φ hφ w hw,
+    rw [linear_map.smul_apply, hφ w hw, smul_zero] } }
+
+@[simp] lemma mem_dual_annihilator (φ : module.dual R M) :
+  φ ∈ W.dual_annihilator ↔ ∀ w ∈ W, φ w = 0 := iff.rfl
+
+/-- The `dual_restrict` of a submodule `W` of `M` is the linear map from the
+  dual of `M` to the dual of `W` such that the domain of each linear map is
+  restricted to `W`. -/
+def dual_restrict (W : submodule R M) :
+  module.dual R M →ₗ[R] module.dual R W :=
+linear_map.dom_restrict' W
+
+@[simp] lemma dual_restrict_apply
+  (W : submodule R M) (φ : module.dual R M) (x : W) :
+  W.dual_restrict φ x = φ x.1 := rfl
+
+lemma dual_restrict_ker_eq_dual_anihilator
+  (W : submodule R M) : W.dual_restrict.ker = W.dual_annihilator :=
+begin
```

Comment:
How awkward does it make things to simply define `dual_annihilator  := W.dual_restrict.ker` so that this is true by definition?

Reply:
Also, typo:
```suggestion
lemma dual_restrict_ker_eq_dual_annihilator
  (W : submodule R M) : W.dual_restrict.ker = W.dual_annihilator :=
begin
```

Reply:
It turns out not to be akward at all since what matters is `mem_dual_annihilator` which we can prove in 7 lines. In total we save about 4 lines but I think changing it to your suggestion is worth it since there are more APIs for kernels I would assume.